### PR TITLE
Fix crash when optimizing protocol composition

### LINKED ISSUE #54035: [SR-11624] Compiler crashes in SILCombiner::propagateConcreteTypeOfInitExistential(swift::FullApplySite)
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-11624      |
   |Radar             | rdar://problem/56373159         |
   |Original Reporter | wooster (JIRA User)      |
   |Type              | Bug    |
   |Status            | Closed        |
   |Resolution        | Done    |
        

Attachment: [Download](https://user-images.githubusercontent.com/2727770/164963497-8d24aaf5-f51f-4456-be3c-abc342bf4c45.gz)

      
<details>
  <summary>Environment</summary>

Xcode 11.1 (11A1027)  
Apple Swift version 5.1 (swiftlang-1100.0.270.13 clang-1100.0.33.7)

 Xcode 11.2 beta 2 (11B44)

Apple Swift version 5.1.1 (swiftlang-1100.0.275.2 clang-1100.0.33.9)

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, CompilerCrash, SILOptimizer, WholeModuleOnly        |
   |Assignee          | @zoecarver      |
   |Priority          | Medium      |

   md5: 60e01b7aaba84d4bacafb1c412c29dd9

  </details>

**Issue Description:**

The compiler crashes when compiling the reduced case:

```swift
class MultiListViewModel<T> {
}

protocol EmptyListViewModel {
}

class Foo<H> {
    init(viewModel: MultiListViewModel<H> & EmptyListViewModel) {
    }
}

final class ItemListViewModel: MultiListViewModel<String> {
}

extension ItemListViewModel: EmptyListViewModel {
}

let item = ItemListViewModel()
let f = Foo(viewModel: item)
```

with the command:

### PR DESCRIPTION
**Pull request by zoecarver**
This patch fixes a compiler crash when trying to optimize a protocol composition between a class and protocol. This is a super simple fix where I simply add a condition to check if the nominal arg type is null. Here's an example program:
```swift
class ClassA<T> { }
protocol ProtocolA { }

class MainClass<H> {
    init(x: ClassA<H> & ProtocolA) { }
}

final class ClassB: ClassA<String> { }
extension ClassB: ProtocolA { }

_ = MainClass(x: ClassB())
```

Resolves #54035

Once this has been reviewed, can someone with commit access trigger swift-ci, please?

### CODE REVIEW

File: validation-test/compiler_crashers_2_fixed/sr11624.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// This source file is part of the Swift.org open source project
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+
+// RUN: not %target-swift-frontend %s -O -whole-module-optimization
```
Comment (CodaFi) [inline validation-test/compiler_crashers_2_fixed/sr11624.swift]: Does WMO make a difference for this test?

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 2]
```diff
@@ -904,7 +904,7 @@ SILCombiner::buildConcreteOpenedExistentialInfoFromSoleConformingType(
         (archetypeTy->getConformsTo().size() == 1)) {
       PD = archetypeTy->getConformsTo()[0];
     } else if (ArgType.isExistentialType() && !ArgType.isAnyObject() &&
-               !SwiftArgType->isAny()) {
+               !SwiftArgType->isAny() && SwiftArgType->getAnyNominal()) {
       PD = dyn_cast<ProtocolDecl>(SwiftArgType->getAnyNominal());
```
Comment (CodaFi) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: This is equivalent to `dyn_cast_or_null` on the next line.  But it sounds like this code needs to handle protocol compositions in a more principled way.  It seems like there’s a missing case here.

File: validation-test/compiler_crashers_2_fixed/sr11624.swift (hunk 1 above)
Comment (zoecarver) [reply validation-test/compiler_crashers_2_fixed/sr11624.swift]: Yes, it only crashes with WMO.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: What missing case is that?

Comment (CodaFi) [reply lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: Not sure @atrick is more familiar with this code.

File: validation-test/compiler_crashers_2_fixed/sr11624.swift (hunk 1 above)
Comment (theblixguy) [reply validation-test/compiler_crashers_2_fixed/sr11624.swift]: Crashes for me with just `-O` as well. Also, I think the above invocation is not correct - it should be `not %target-swift-frontend -O -whole-module-optimization %s -emit-sil`.

**Review by atrick — APPROVED** [review]
LGTM. @CodaFi is right, `dyn_cast_or_null` is cleaner.

There are various scenarios related to protocol composition that we should be able to optimize. This optimization is not really equipped to handle them though so it's best to just bail here.

I filed a bug for these issues:
SR-11695: SILOptimizer should devirtualize calls to a protocol composition type

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
Review comments addressed in #28032

**Comment by zoecarver** [comment]
@atrick, thank you for the review and for fixing the review comments!

**Pull request merged**